### PR TITLE
Unhardcode monster reach attacks, fix some bugs

### PR DESCRIPTION
**Pull request by Venera3**
#### Summary
None

#### Purpose of change

A lot of monster special attacks are hardcoded, even though there's a very nice JSON way of defining them (including custom messages, so that e.g. dragonfly larvae don't stretch their arms at you at range). This PR adds the functionality of defining reaching melee attacks to it, and migrates the existing attacks over to use it. While testing, I fixed some wonkiness as well - unless the monster's special attack entry specified some fields the code would default to the fallback value instead of reading it from the actual attack definition.

#### Describe the solution

Added some new fields to mattack_actor:
- `range`, allowing you to define reach attacks using it,
- `no_adjacent`, disqualifying targeting an adjacent enemy with an attack,
- `attack_chance`, allowing for a special attack to fail even if it got chosen in the initial attack roll. This doesn't cost moves or resets the cooldown for the monster, but it does preclude it from using any specials that round.

Added a JSON equivalent to ~almost~ all hardcoded reach attacks. 
Updated mattack_actor to use `optional`, fixing the abovementioned wonkiness.

TODO:
- [x] Differentiate `LONGSWIPE` into a reach attack and a throat-cutting one
- [x] Update the docs
- [x] Test it a bit more

#### Describe alternatives you've considered

~Also adding a field to disqualify the attacks from targeting adjacent targets.~ Turns out this was a part of `LONGSWIPE`.
Adjust some damage numbers/balance some attacks (Skellyjuggs doing ~9 dmg per hit, really?), but combat testing takes time and this opens up the attacks for other contributors to mess around with.
Include a check for Uncanny dodge - the hardcoded attacks use explicit checks for it instead of going through the normal melee dodge calc. I think these interactions should become part of the base dodge calculations instead of only affecting special attacks.

#### Testing

Attacks with a defined range will target the player if there's a clear path equivalent to their range, won't trigger through glass.
`Min/max_mul`, `move_cost` and `accuracy` of the base attack definition carry over properly, and can be overwritten by the monster's own special attack data. Nonadjacent attacks don't target adjacent hostiles (they aren't smart enough to switch targets, but that's a project for somebody competent). Failed `attack_chance` rolls preclude the attack from triggering, don't allow for other attacks on the same turn.
With 3d vision on they can attack upwards, but not through obstacles (standing on th edge of the roof is dangerous, a buffer of 2 tiles is safe).

#### Additional context

For now I just copied over most attack data from the existing hardcoded attacks. Still, some balance changes do result from this: flesh raptors got a slight nerf in that they can't attack and impale you on the same turn, slam and scratch both got slightly slower (`move_cost` defaulted to 100), and Uncanny dodge becomes slightly more useless - see alternatives.
It also cuts some very edge-case interactions, like hollow zombie's head extending towards you, going *thud* on a window and them getting stunned for 6 turns. Funny, yes, but not easily genericizable.

### CODE REVIEW

File: data/json/monster_special_attacks/monster_attacks.json [hunk 1]
```diff
@@ -13,6 +13,53 @@
...
+    "cooldown": 20,
+    "move_cost": 200,
+    "range": 3,
+    "damage_max_instance":[ { "damage_type": "stab", "amount": 10 } ],
+    "hit_dmg_u": "The %1$s thrusts its arms at you, and its claws pierce your %2$s!",
+    "hit_dmg_npc": "The %1$s thrusts its arms at <npcname>, and its claws pierce their %2$s!",
+    "no_dmg_msg_u": "The %1$s's arms hit your %2$s, but they glance off your armor!",
+    "no_dmg_msg_npc": "The %1$s's arms hit <npcname>'s %2$s, but they glance off their armor!",
+    "miss_msg_u": "The %s thrusts its arms at you, but you evade the stretched arm!",
+    "miss_msg_npc": "The %s's thrustst its arms at <npcname>, but they dodge!"
```
Comment (actual-nh) [inline data/json/monster_special_attacks/monster_attacks.json:61]:
```suggestion
    "miss_msg_npc": "The %s's thrusts its arms at <npcname>, but they dodge!"
```

File: data/json/monsters/insect_spider.json [hunk 2]
```diff
@@ -595,7 +595,18 @@
...
+    "special_attacks": [ 
+      {
+        "id": "stretch_attack",
+        "range": 2,
+        "hit_dmg_u": "The %1$s's mouthpart shoots out at you, and it pierces your %2$s!",
+        "hit_dmg_npc": "The %1$s's mouthpart shoots out at <npcname>, and it pierces their %2$s!",
+        "no_dmg_msg_u": "The %1$s's mouthpart hits your %2$s, but it glances off your armor!",
+        "no_dmg_msg_npc": "The %1$s's mouthpart hits <npcname>'s %2$s, but it glances off their armor!",
+        "miss_msg_u": "The %s thrusts its mouthpart at you, but you dodge!",
+        "miss_msg_npc": "The %s's thrustst its mouthpart at <npcname>, but they dodge!"
```
Comment (actual-nh) [inline data/json/monsters/insect_spider.json:607]:
```suggestion
        "miss_msg_npc": "The %s's thrusts its mouthpart at <npcname>, but they dodge!"
```

**Review by actual-nh — COMMENTED** [review]
Interesting!

**Comment by PatrikLundell** [comment]
Should the range attack check really be for a path of a sufficiently short range? I would expect all normal cases to require a straight path and that the attack wouldn't be available if it would have to path around obstacles. I'd also expect ranged attacks to come into the categories of requiring a pathable path (i.e. you can walk along it) and an open path (i.e. there's nothing in the way, but you may not be able to move along it, such as e.g. from one roof top to another or otherwise through the air, over rubble, or through an open window). The latter category would essentially be the same as the criteria for a missile weapon (minus the ability to attack through obstacles by destroying them).

**Comment by Venera3** [comment]
`map::clear_path` is the same function that the ballistics code uses, actually. It paths over empty air without a problem, and searches for a straight line with a max offset of one tile in every direction. I'd *like* for there to be a variant that uses proper pathfinding (for big scary tentacle monsters and the like), but at a cursory glance there is no function that does that that I could repurpose easily - I'd be very happy to be corrected there, of course!

**Pull request merged**